homeassistant/components/control4/config_flow.py

```diff
@@ -0,0 +1,205 @@
+    }
+
+
+class Control4Validator:
+    """Validates that config details can be used to authenticate and communicate with Control4."""
+
+    def __init__(self, host, username, password):
+        """Initialize."""
+        self.host = host
+        self.username = username
+        self.password = password
+        self.account = C4Account(self.username, self.password)
+        self.controller_name = None
+        self.director_bearer_token = None
+        self.director = None
+
+    async def authenticate(self) -> bool:
+        """Test if we can authenticate with the Control4 account API."""
+        try:
+            # Authenticate with Control4 account
+            await self.account.getAccountBearerToken()
+
+            # Get controller name
+            account_controllers = await self.account.getAccountControllers()
+            self.controller_name = account_controllers["controllerCommonName"]
+
+            # Get bearer token to communicate with controller locally
+            self.director_bearer_token = await self.account.getDirectorBearerToken(
+                self.controller_name
+            )
+            return True
+        except (BadCredentials, Unauthorized) as exception:
+            _LOGGER.error(exception)
+            return False
+
+    async def connect_to_director(self) -> bool:
+        """Test if we can connect to the local Control4 Director."""
+        try:
+            self.director = C4Director(self.host, self.director_bearer_token)
+            await self.director.getAllItemInfo()
+            return True
+        except Exception as exception:  # pylint: disable=broad-except
+            _LOGGER.error(exception)
+            return False
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+
+    hub = Control4Validator(data["host"], data["username"], data["password"])
+
+    if not await hub.authenticate():
+        raise InvalidAuth
+
+    if not await hub.connect_to_director():
+        raise CannotConnect
+
+    # If you cannot connect:
+    # throw CannotConnect
+    # If the authentication is wrong:
+    # InvalidAuth
+
+    # Return info that you want to store in the config entry.
+    return {
+        "controller_name": hub.controller_name,
+    }
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Control4."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+
+            try:
+                info = await validate_input(self.hass, user_input)
+                if info["controller_name"] in configured_instances(self.hass):
+                    return self.async_abort(reason="already_configured")
```

Comment:
Abort inside the try handler will not abort because it will be caught by the broad exception